src/ClientMessage.ts

```diff
@@ -15,248 +15,222 @@
     }
 
-    hasFlags(flags: number): number {
-        return this.getFlags() & flags;
+    next(): Frame {
+        const result = this.nextFrame;
+        if (this.nextFrame != null) {
+            this.nextFrame = this.nextFrame.next;
+        }
+        return result;
     }
 
-    getFrameLength(): number {
-        return this.buffer.readInt32LE(BitsUtil.FRAME_LENGTH_FIELD_OFFSET);
+    hasNext(): boolean {
+        return this.nextFrame !== null;
     }
 
-    setFrameLength(value: number): void {
-        this.buffer.writeInt32LE(value, BitsUtil.FRAME_LENGTH_FIELD_OFFSET);
+    peekNext(): Frame {
+        return this.nextFrame;
     }
+}
 
-    getDataOffset(): number {
-        return this.buffer.readInt16LE(BitsUtil.DATA_OFFSET_FIELD_OFFSET);
-    }
+export class ClientMessage {
+    startFrame: Frame;
+    endFrame: Frame;
+    private retryable: boolean;
+    private connection: ClientConnection;
 
-    setDataOffset(value: number): void {
-        this.buffer.writeInt16LE(value, BitsUtil.DATA_OFFSET_FIELD_OFFSET);
+    private constructor(startFrame?: Frame, endFrame?: Frame) {
+        this.startFrame = startFrame;
+        this.endFrame = endFrame || startFrame;
     }
 
-    setRetryable(value: boolean): void {
-        this.isRetryable = value;
+    static createForEncode(): ClientMessage {
+        return new ClientMessage();
     }
 
-    appendByte(value: number): void {
-        this.buffer.writeUInt8(value, this.cursor);
-        this.cursor += BitsUtil.BYTE_SIZE_IN_BYTES;
+    static createForDecode(startFrame: Frame): ClientMessage {
+        return new ClientMessage(startFrame);
     }
 
-    appendBoolean(value: boolean): void {
-        return this.appendByte(value ? 1 : 0);
+    static isFlagSet(flags: number, flagMask: number): boolean {
+        const i = flags & flagMask;
+        return i === flagMask;
     }
 
-    appendInt32(value: number): void {
-        this.buffer.writeInt32LE(value, this.cursor);
-        this.cursor += BitsUtil.INT_SIZE_IN_BYTES;
+    getStartFrame(): Frame {
+        return this.startFrame;
     }
 
-    appendUint8(value: number): void {
-        this.buffer.writeUInt8(value, this.cursor);
-        this.cursor += BitsUtil.BYTE_SIZE_IN_BYTES;
-    }
+    add(frame: Frame): void {
+        frame.next = null;
+        if (this.startFrame == null) {
+            this.startFrame = frame;
+            this.endFrame = frame;
+            return;
+        }
 
-    appendLong(value: any): void {
-        this.writeLongInternal(value, this.cursor);
-        this.cursor += BitsUtil.LONG_SIZE_IN_BYTES;
+        this.endFrame.next = frame;
+        this.endFrame = frame;
     }
 
-    appendString(value: string): void {
-        const length = Buffer.byteLength(value, 'utf8');
-        this.buffer.writeInt32LE(length, this.cursor);
-        this.cursor += 4;
-        this.buffer.write(value, this.cursor);
-        this.cursor += length;
+    frameIterator(): ForwardFrameIterator {
```

Comment:
Great idea. Changed the codecs as you suggested

Reply:
Do you think we should keep `frameIterator()` method on `ClientMessage`? I'd rather initialize the `nextFrame` field in the constructor and, maybe, add `resetIterator()` method which would do `this.nextFrame = this.startFrame` (if we need reset functionality, of course). WDYT?

Another thing to mention here is that I'd leave `nextFrame()` method name instead of the `getNextFrame()` one. It makes the fact that it's not a dump getter method a bit more obvious. But that's a nit for sure.